JWBestPracticeApps/JWConviva/AnalyticsObserver.swift

```diff
@@ -119,24 +123,22 @@ class AnalyticsObserver: NSObject {
     }
     
     func cleanupVideoPsm() {
-        if (client != nil && psmVideoInstance != nil) {
-            client!.releasePlayerStateManager(psmVideoInstance)
-            psmVideoInstance = nil
-        }
+        guard let client = self.client, let psmVideoInstance = self.psmVideoInstance else { return }
+        client.releasePlayerStateManager(psmVideoInstance)
+        self.psmVideoInstance = nil
     }
     
     func attachVideoPlayer() {
-        if (client != nil) {
-            
-            if (psmVideoInstance == nil) {
-                psmVideoInstance = client!.getPlayerStateManager()
-                psmVideoInstance?.setPlayerVersion!(JWPlayerController.sdkVersion())
-                psmVideoInstance?.setPlayerType!("JWPlayer")
-            }
+        guard let client = self.client else { return }
             
-            if (psmVideoInstance != nil && videoSessionID != NO_SESSION_KEY && !client!.isPlayerAttached(videoSessionID)){
-                client!.attachPlayer(videoSessionID, playerStateManager: psmVideoInstance)
-            }
+        if (psmVideoInstance == nil) {
+            psmVideoInstance = client.getPlayerStateManager()
+            psmVideoInstance?.setPlayerVersion!(JWPlayerController.sdkVersion())
```

Comment:
this puzzles me, why does it require a '!' at the end of the method name? the method was declared as:
- (void)setPlayerVersion:(NSString *)plVer;
Can you look at all the psmVideoInstance code?

Reply:
The class is a Protocol and setPlayerVersion is optional to implement.